src/coprocessor/endpoint.rs

```diff
@@ -11,641 +11,448 @@
+                                handler.collect_metrics_into(&mut metrics);
+                                metrics
+                            };
+
+                            tracker.on_finish_item(Some(exec_metrics));
+                            let exec_details = tracker.get_item_exec_details();
+
+                            let (mut resp, finished) = match result {
+                                Err(e) => (make_error_response(e), true),
+                                Ok((None, _)) => {
+                                    let yielded = Either::Right(tracker);
+                                    let next_state = None;
+                                    return Some(Ok((yielded, next_state)));
+                                }
+                                Ok((Some(resp), finished)) => (resp, finished),
+                            };
+                            resp.set_exec_details(exec_details);
+
+                            let yielded = Either::Left(resp);
+                            let next_state = Some((tracker, handler, finished));
+                            Some(Ok((yielded, next_state)))
+                        }
+                        None => {
+                            // Finished
+                            None
+                        }
                     }
-                    let key = req.get_request_key();
-                    grouped_reqs.entry(key).or_insert_with(Vec::new).push(req);
-                }
-                Task::SnapRes(q_id, snap_res) => {
-                    self.handle_snapshot_result(q_id, snap_res);
-                }
-            }
-        }
+                }).filter_map(|resp_or_tracker| match resp_or_tracker {
+                    Either::Left(resp) => Some(resp),
+                    Either::Right(mut tracker) => {
+                        tracker.on_finish_all_items();
+                        None
+                    }
+                })
+            })
+            .flatten()
+    }
 
-        for (_, mut reqs) in grouped_reqs {
-            let max_running_task_count = self.max_running_task_count;
-            if self.running_task_count() >= max_running_task_count {
-                self.notify_failed(Error::Full(max_running_task_count), reqs);
-                continue;
-            }
-            for req in &mut reqs {
-                let task_count = Arc::clone(&self.running_task_count);
-                req.tracker.task_count(task_count);
+    /// Handle a stream request and run on the read pool. Returns a stream producing each
+    /// result, which must be a `Response` and will never fail. If there are errors during
+    /// handling, they will be embedded in the `Response`.
+    fn handle_stream_request(
+        &self,
+        req_ctx: ReqContext,
+        handler_builder: RequestHandlerBuilder<E::Snap>,
+    ) -> impl Stream<Item = coppb::Response, Error = ()> {
+        let (tx, rx) = mpsc::channel::<coppb::Response>(self.stream_channel_size);
```

Comment:
How about use `util::mpsc::channel`? According to jay's bench, it's much better than futures::Sender.

Reply:
Ok, I will have a test